src/Compilers/Core/MSBuildTask/Csc.cs

```diff
@@ -202,11 +202,14 @@ protected internal override void AddResponseFileCommands(CommandLineBuilderExten
 
             // If not design time build and the globalSessionGuid property was set then add a -globalsessionguid:<guid>
             bool designTime = false;
-            if (HostObject != null)
+            if (HostObject is ICscHostObject csHost)
             {
-                var csHost = HostObject as ICscHostObject;
                 designTime = csHost.IsDesignTime();
             }
+            else if (HostObject != null)
+            {
+                throw new InvalidOperationException();
```

Comment:
I think ideally these should all have messages with them. The only exception which doesn't require a message is items like UnreachableException. 

As for localization yeah they need to be localized. You can put the message in ErrorStrings.resx though, use the ID and everything else will be handled.